src/scorepy/events.cpp

```diff
@@ -52,45 +38,119 @@ static const std::array<std::string, 2> EXIT_REGION_WHITELIST = {
 #endif
 };
 
-void region_end(const std::string& region_name)
+void region_begin(const std::string& region, const std::string& module,
+                  const std::string& file_name, const std::uint64_t line_number,
+                  const std::uintptr_t& identifier)
 {
-    const auto itRegion = regions.find(region_name);
-    if (itRegion != regions.end())
-    {
-        SCOREP_User_RegionEnd(itRegion->second.value);
-    }
-    else
+    auto& region_handle = regions[identifier];
+
+    if (region_handle == uninitialised_region_handle)
     {
-        static region_handle error_region;
-        static SCOREP_User_ParameterHandle scorep_param = SCOREP_USER_INVALID_PARAMETER;
-        static bool error_printed = false;
+        const auto& region_name = make_region_name(module, region);
+        auto it = user_regions.find(region_name);
+        if (it == user_regions.end())
+        {
+            SCOREP_User_RegionInit(&region_handle.value, NULL, &SCOREP_User_LastFileHandle,
```

Comment:
Not introduced here but as this has been changed (and even copied): Passing `&SCOREP_User_LastFileHandle` is not required at least and wrong at worst. It should be `NULL`.

Reason: Score-P assumes that the file_name is a string constant and has some optimization that reuses the file handle from the previous call if the `const char*` of the file_name matches the one from the previous call. As we use dynamically allocated memory here this would be  a bad idea.

However: As the 2nd parameter is NULL this isn't currently being done and the 3rd parameter is simply ignored. For completeness I'd still change it

Reply:
I remember having this discussion with Bert. I am not sure, why this is still as it is. Maybe becaus its currently not an issue. I'll change this as suggested.

Reply:
For reference the relevant code in Score-P is:
```
    if ( lastFileName && lastFile && *lastFileName == file )
    {
        SCOREP_MutexUnlock( scorep_user_file_table_mutex );
        return *lastFile;
    }
```
lastFileName and lastFile are the 2nd and 3rd parameter in the call. So yes it is currently not an issue (see also my last sentence in the previous comment)

Reply:
done